raft/src/main/java/org/apache/kafka/snapshot/FileRawSnapshotWriter.java

```diff
@@ -53,13 +54,13 @@ public long sizeInBytes() throws IOException {
     }
 
     @Override
-    public void append(ByteBuffer buffer) throws IOException {
+    public void append(MemoryRecords records) throws IOException {
```

Comment:
I am considering how to improve this API in the interim to make the alignment expectation clearer. In fact, the `BaseRecords` interface is really minimal and makes no assumptions about alignment. Suppose we introduce two new types:

```java
// Represents a record set which is not necessarily offset-aligned
interface UnalignedRecords extends BaseRecords;

class UnalignedMemoryRecords extends UnalignedRecords {
  ByteBuffer buffer;
}
class UnalignedFileRecords extends UnalignedRecords {
  FileChannel channel;
}
```

I think these could be very simple implementations and they would improve the type safety of these APIs. Then we can use `UnalignedRecords` in the in the `RawSnapshotWriter` and `RawSnapshotReader` interfaces which restores the symmetry. 

What do you think?

Reply:
This will be more clear and got 3 little problems when I tried:
1. `Readable.readRecords` still returns a `MemoryRecords` and we need to convert it to `UnalignedMemoryRecords`, this is simple
2. We need to implement `UnalignedFileRecords.toSend` which need to return RecordSend, but this is unreasonable, I added a `FileChannelSend` similar to `ByteBufferSend` to solve this problem
3. We should still use `UnalignedMemoryRecords` in `FileRawSnapshotWriter` so we don't restore symmetry

@hachikuji @jsancio ,I think my implementation is not the perfect one, please take a look to see whether we could make some improvement, thank you.